padl/dumptools/sourceget.py

```diff
@@ -28,11 +28,20 @@ def get_source(filename: str, use_replace_cache: bool = True) -> str:
     If *use_replace_cache*, try getting the source from the "replace_cache", which can contain
     explicit replacements of the original source strings.
     """
+    # needed to support doctest
+    if filename.startswith('<doctest'):
```

Comment:
Nice! Great to see that `doctest` had a solution.

Reply:
@wuhu Could you point me to a good resource for learning about these different caching mechanisms in `python`? I've had a hard time finding a good explanation for how to navigate these different cases.

Reply:
I don't have one but it's this:
The frameinfo (see https://docs.python.org/3/library/inspect.html) object has a filename field. This allows us to get the source of what's currently being executed.
For a normal python module it's the filename.
For ipython and doctest the code is not in files, therefore it's different.
For ipython, every cell has an id and the "filename" contains a string with that id. The content of the code can than be found in linecache.cache (https://docs.python.org/3/library/linecache.html).
For doctest neither works, the code in the docstrings is just executed. We can get the code from the fileinfo.code_context which the context of the code being executed - this is not perfect as it contains only a few lines, but for doctest, it looks like it's enough.